Add GPU implementation for Range op

benbarsdell opened the pull request:
- This op currently has a GPU kernel registered, but it computes and returns its output on the host, which can cause significant slowdowns due to H2D memcopies and slow CPU execution. I don't know if there is a special reason it's like this (hopefully not).
- This patch adds a proper GPU kernel that returns the output on the device.

cc @nluehr

benbarsdell committed:
Add GPU implementation for Range op

- This op currently has a GPU kernel registered, but it computes and returns its output on the host,
  which can cause significant slowdowns due to H2D memcopies and slow CPU execution.
- This patch adds a proper GPU kernel and returns the output on the device.

kkimdev commented:
cc @JXRiver @bgdax 

Update: we'll be evaluating if it's OK to register (e.g., no perf regression) Range GPU kernel.

JXRiver commented:
Thanks for your contribution. @bgdax is performing some benchmark tests for this PR. The initial result looks promising. We need some more time to thoroughly quantify its performance impact. Will get back to you once that is done.

bgdax commented on tensorflow/core/kernels/sequence_ops.cc:
```diff
@@ -82,27 +102,23 @@ class RangeOp : public OpKernel {
   REGISTER_KERNEL_BUILDER(Name("Range")                      \
                               .Device(DEV)                   \
                               .HostMemory("start")           \
                               .HostMemory("limit")           \
```
Thank you for the contribution!

Currently there are many ops that require inputs and/or outputs in host memory for int32.
As a result, in the current state, it leads to performance degradation in our benchmarks in some int32 use-cases due to the extra D2H copies.

I suggest using the existing implementation for the int32 case, with output in host memory.

One way to achieve this would be to define an additional registration macro here (e.g. REGISTER_HOST_OUTPUT_KERNEL), use it in REGISTER_CPU_KERNEL below, and remove the 'TF_CALL_int32(REGISTER_GPU_KERNEL)' registration. 
The CPU implementation with output in host memory will then be used as the fallback for int32.

benbarsdell replied:
Done

bgdax reviewed (changes requested):
Please see the comment on the int32 registration.

benbarsdell committed:
Execute int32 GPU Range kernel on the CPU instead

- This is to avoid performance regressions due to device->host copies,
  as most int32 tensors are expected in host memory.

rohan100jain reviewed (commented):
Can we add a test? You can check the tensor.backing_device for some int64 computation.

gbaned commented:
@benbarsdell Can you please check @rohan100jain's comments and keep us posted ? Thanks!

benbarsdell commented:
> Can we add a test? You can check the tensor.backing_device for some int64 computation.

I'm not sure I understand the goal of the test, could you elaborate?

JXRiver commented:
My understanding of @rohan100jain's comment is that we want to make sure int32 is forced to be placed on HostMemory and int64 could be placed in HostMemory or DeviceMemory. So future developers won't accidentally change this behavior. For example:
```
tf.config.set_soft_device_placement(False)
with tf.device('/GPU:0'):
  # int32 is forced to be placed on HostMemory to avoid unnecessary D2H copies
  t = tf.range(10, dtype=tf.int32)
  print(t.backing_device)  # assert CPU

  t = tf.range(10, dtype=tf.int64)
  print(t.backing_device)  # assert GPU

with tf.device('/CPU:0'):
  t = tf.range(10, dtype=tf.int32)
  print(t.backing_device)  # assert CPU

  t = tf.range(10, dtype=tf.int64)
  print(t.backing_device)  # assert CPU
```

You may add tests to https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/ops/math_ops_test.py or create a file in https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/kernel_tests/math_ops
